The job **`dist-x86_64-linux-alt`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17468345866/job/49610036779) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49610036779) [(plain)](https://github.com/rust-lang/rust/commit/30e8aea19f1a447e0575eafbdd50b95799c6de1b/checks/49610036779/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking snapbox v0.6.21
error[E0725]: the feature `specialization` is not in the list of allowed features
   --> /rust/deps/im-rc-15.1.0/src/lib.rs:341:41
    |
341 | #![cfg_attr(has_specialisation, feature(specialization))]
    |                                         ^^^^^^^^^^^^^^

error[E0658]: specialization is unstable
   --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:111:5
    |
111 | /     default fn search_key<BK>(slice: &[Self], key: &BK) -> Result<usize, usize>
112 | |     where
113 | |         BK: Ord + ?Sized,
114 | |         Self::Key: Borrow<BK>,
115 | |     {
116 | |         slice.binary_search_by(|value| Self::Key::borrow(&value.0).cmp(key))
117 | |     }
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:119:5
    |
119 | /     default fn search_value(slice: &[Self], key: &Self) -> Result<usize, usize> {
120 | |         slice.binary_search_by(|value| value.0.cmp(&key.0))
121 | |     }
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
    --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:1693:5
     |
1693 | /     default fn eq(&self, other: &Self) -> bool {
1694 | |         self.len() == other.len() && self.diff(other).next().is_none()
1695 | |     }
     | |_____^
     |
     = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
     = help: add `#![feature(specialization)]` to the crate attributes to enable
     = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /rust/deps/im-rc-15.1.0/src/ord/set.rs:119:5
    |
119 | /     default fn search_key<BK>(slice: &[Self], key: &BK) -> Result<usize, usize>
120 | |     where
121 | |         BK: Ord + ?Sized,
122 | |         Self::Key: Borrow<BK>,
123 | |     {
124 | |         slice.binary_search_by(|value| Self::Key::borrow(value).cmp(key))
125 | |     }
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /rust/deps/im-rc-15.1.0/src/ord/set.rs:127:5
    |
127 | /     default fn search_value(slice: &[Self], key: &Self) -> Result<usize, usize> {
128 | |         slice.binary_search_by(|value| value.cmp(key))
129 | |     }
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
    --> /rust/deps/im-rc-15.1.0/src/hash/map.rs:1583:5
     |
1583 | /     default fn eq(&self, other: &Self) -> bool {
1584 | |         self.test_eq(other)
1585 | |     }
     | |_____^
     |
     = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
     = help: add `#![feature(specialization)]` to the crate attributes to enable
     = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
    --> /rust/deps/im-rc-15.1.0/src/hash/map.rs:1783:5
     |
1783 | /     default fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
1784 | |         let mut d = f.debug_map();
1785 | |         for (k, v) in self {
1786 | |             d.entry(k, v);
1787 | |         }
1788 | |         d.finish()
1789 | |     }
     | |_____^
     |
     = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
     = help: add `#![feature(specialization)]` to the crate attributes to enable
     = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
   --> /rust/deps/im-rc-15.1.0/src/hash/set.rs:823:5
    |
823 | /     default fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
824 | |         f.debug_set().entries(self.iter()).finish()
825 | |     }
    | |_____^
    |
    = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
    = help: add `#![feature(specialization)]` to the crate attributes to enable
    = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0658]: specialization is unstable
    --> /rust/deps/im-rc-15.1.0/src/vector/mod.rs:1724:5
     |
1724 | /     default fn eq(&self, other: &Self) -> bool {
1725 | |         self.len() == other.len() && self.iter().eq(other.iter())
1726 | |     }
     | |_____^
     |
     = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
     = help: add `#![feature(specialization)]` to the crate attributes to enable
     = note: this compiler was built on 2025-09-04; consider upgrading it if it is out of date

error[E0119]: conflicting implementations of trait `PartialEq` for type `OrdMap<_, _>`
    --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:1699:1
     |
1688 | / impl<K, V> PartialEq for OrdMap<K, V>
1689 | | where
1690 | |     K: Ord + PartialEq,
1691 | |     V: PartialEq,
     | |_________________- first implementation here
...
1699 | / impl<K, V> PartialEq for OrdMap<K, V>
1700 | | where
1701 | |     K: Ord + Eq,
1702 | |     V: Eq,
     | |__________^ conflicting implementation for `OrdMap<_, _>`

error[E0119]: conflicting implementations of trait `PartialEq` for type `hash::map::HashMap<_, _, _>`
    --> /rust/deps/im-rc-15.1.0/src/hash/map.rs:1589:1
     |
1577 | / impl<K, V, S> PartialEq for HashMap<K, V, S>
---

error[E0119]: conflicting implementations of trait `PartialEq` for type `Vector<_>`
    --> /rust/deps/im-rc-15.1.0/src/vector/mod.rs:1730:1
     |
1723 | impl<A: Clone + PartialEq> PartialEq for Vector<A> {
     | -------------------------------------------------- first implementation here
...
1730 | impl<A: Clone + Eq> PartialEq for Vector<A> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Vector<_>`

error[E0119]: conflicting implementations of trait `Debug` for type `hash::map::HashMap<_, _, _>`
    --> /rust/deps/im-rc-15.1.0/src/hash/map.rs:1793:1
     |
1777 | / impl<K, V, S> Debug for HashMap<K, V, S>
1778 | | where
1779 | |     K: Hash + Eq + Debug,
1780 | |     V: Debug,
1781 | |     S: BuildHasher,
     | |___________________- first implementation here
...
1793 | / impl<K, V, S> Debug for HashMap<K, V, S>
1794 | | where
1795 | |     K: Hash + Eq + Ord + Debug,
1796 | |     V: Debug,
1797 | |     S: BuildHasher,
     | |___________________^ conflicting implementation for `hash::map::HashMap<_, _, _>`

error[E0119]: conflicting implementations of trait `Debug` for type `hash::set::HashSet<_, _>`
   --> /rust/deps/im-rc-15.1.0/src/hash/set.rs:829:1
    |
818 | / impl<A, S> Debug for HashSet<A, S>
819 | | where
820 | |     A: Hash + Eq + Debug,
821 | |     S: BuildHasher,
    | |___________________- first implementation here
...
829 | / impl<A, S> Debug for HashSet<A, S>
830 | | where
831 | |     A: Hash + Eq + Debug + Ord,
832 | |     S: BuildHasher,
    | |___________________^ conflicting implementation for `hash::set::HashSet<_, _>`

error[E0119]: conflicting implementations of trait `BTreeValue` for type `(_, _)`
   --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:137:1
    |
104 | impl<K: Ord, V> BTreeValue for (K, V) {
    | ------------------------------------- first implementation here
...
137 | impl<K: Ord + Copy, V> BTreeValue for (K, V) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `(_, _)`

error[E0119]: conflicting implementations of trait `BTreeValue` for type `ord::set::Value<_>`
   --> /rust/deps/im-rc-15.1.0/src/ord/set.rs:145:1
    |
112 | impl<A: Ord> BTreeValue for Value<A> {
    | ------------------------------------ first implementation here
...
145 | impl<A: Ord + Copy> BTreeValue for Value<A> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `ord::set::Value<_>`

[RUSTC-TIMING] blake3 test:false 0.724
    Checking handlebars v6.3.1
error[E0308]: mismatched types
   --> /rust/deps/im-rc-15.1.0/src/ord/map.rs:143:73
    |
137 | impl<K: Ord + Copy, V> BTreeValue for (K, V) {
    |      - expected type parameter
138 |     fn search_key<BK>(slice: &[Self], key: &BK) -> Result<usize, usize>
    |                   -- found type parameter
...
143 |         linear_search_by(slice, |value| Self::Key::borrow(&value.0).cmp(key))
    |                                                                     --- ^^^ expected `&K`, found `&BK`
    |                                                                     |
    |                                                                     arguments to this method are incorrect
    |
    = note: expected reference `&K`
               found reference `&BK`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
note: method defined here
   --> /rustc/30e8aea19f1a447e0575eafbdd50b95799c6de1b/library/core/src/cmp.rs:979:8
    |
979 |     fn cmp(&self, other: &Self) -> Ordering;
    |        ^^^

[RUSTC-TIMING] clap_complete test:false 1.356
    Checking tracing-chrome v0.7.2
error[E0308]: mismatched types
   --> /rust/deps/im-rc-15.1.0/src/ord/set.rs:151:70
    |
146 |     fn search_key<BK>(slice: &[Self], key: &BK) -> Result<usize, usize>
    |                   -- found this type parameter
...
151 |         linear_search_by(slice, |value| Self::Key::borrow(value).cmp(key))
    |                                                                  --- ^^^ expected `&Value<A>`, found `&BK`
    |                                                                  |
    |                                                                  arguments to this method are incorrect
    |
    = note: expected reference `&ord::set::Value<A>`
               found reference `&BK`
note: method defined here
   --> /rustc/30e8aea19f1a447e0575eafbdd50b95799c6de1b/library/core/src/cmp.rs:979:8
    |
979 |     fn cmp(&self, other: &Self) -> Ordering;
    |        ^^^

[RUSTC-TIMING] tracing_chrome test:false 0.265
    Checking rand v0.9.2
[RUSTC-TIMING] snapbox test:false 1.533
---
[RUSTC-TIMING] pasetors test:false 1.058
[RUSTC-TIMING] rustfix test:false 0.672
[RUSTC-TIMING] handlebars test:false 2.709
[RUSTC-TIMING] cargo_util_schemas test:false 2.869
Bootstrap failed while executing `dist --host x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu --include-default-paths build-manifest bootstrap`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo doc --target x86_64-unknown-linux-gnu -Zbinary-dep-depinfo -j 8 -Zroot-dir=/checkout --locked --color always --release --manifest-path /checkout/src/tools/cargo/Cargo.toml --features all-static -Zskip-rustdoc-fingerprint --no-deps -p cargo -p cargo-credential -p cargo-platform -p cargo-test-macro -p cargo-test-support -p cargo-util -p cargo-util-schemas -p crates-io -p mdman -p rustfix [workdir=/checkout]` failed with exit code 101
Created at: src/bootstrap/src/core/build_steps/tool.rs:189:21
Executed at: src/bootstrap/src/core/build_steps/doc.rs:1122:1

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 1:06:58
  local time: Thu Sep  4 16:24:31 UTC 2025
  network time: Thu, 04 Sep 2025 16:24:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/bootstrap/src/core/build_steps/doc.rs
+++ b/src/bootstrap/src/core/build_steps/doc.rs
@@ -965,9 +965,11 @@ macro_rules! tool_doc {
     (
         $tool: ident,
         $path: literal,
-        mode = $mode:expr,
-        $(is_library = $is_library:expr,)?
-        $(crates = $crates:expr)?
+        mode = $mode:expr
+        $(, is_library = $is_library:expr )?
+        $(, crates = $crates:expr )?
+        // Subset of nightly features that are allowed to be used when documenting
+        $(, allow_features: $allow_features:expr )?
        ) => {
         #[derive(Debug, Clone, Hash, PartialEq, Eq)]
         pub struct $tool {
@@ -1041,6 +1043,15 @@ macro_rules! tool_doc {
                     source_type,
                     &[],
                 );
+                let allow_features = {
+                    let mut _value = "";
+                    $( _value = $allow_features; )?
+                    _value
+                };
+
+                if allow_features.is_empty() {
+                    cargo.allow_features(allow_features);
+                }
 
                 cargo.arg("-Zskip-rustdoc-fingerprint");
                 // Only include compiler crates, no dependencies of those, such as `libc`.
@@ -1134,7 +1145,8 @@ tool_doc!(
         "crates-io",
         "mdman",
         "rustfix",
-    ]
+    ],
+    allow_features: "specialization"
 );
 tool_doc!(Tidy, "src/tools/tidy", mode = Mode::ToolBootstrap, crates = ["tidy"]);
 tool_doc!(

--- a/src/bootstrap/src/core/build_steps/doc.rs
+++ b/src/bootstrap/src/core/build_steps/doc.rs
@@ -1005,7 +1005,7 @@ macro_rules! tool_doc {
                     }
                     Mode::ToolTarget => {
                         // target tools should be documented with the in-tree compiler
-                        prepare_doc_compiler(run.builder, run.builder.host_target, 2)
+                        prepare_doc_compiler(run.builder, run.builder.host_target, run.builder.top_stage)
                     }
                     _ => {
                         panic!("Unexpected tool mode for documenting: {:?}", $mode);

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -1131,6 +1131,59 @@ mod snapshot {
     }
 
     #[test]
+    fn dist_compiler_docs() {
+        let ctx = TestCtx::new();
+        insta::assert_snapshot!(
+            ctx.config("dist")
+                .path("rustc-docs")
+                .args(&["--set", "build.compiler-docs=true"])
+                .render_steps(), @r"
+        [build] rustc 0 <host> -> UnstableBookGen 1 <host>
+        [build] rustc 0 <host> -> Rustbook 1 <host>
+        [doc] unstable-book (book) <host>
+        [build] llvm <host>
+        [build] rustc 0 <host> -> rustc 1 <host>
+        [build] rustc 1 <host> -> std 1 <host>
+        [doc] book (book) <host>
+        [doc] book/first-edition (book) <host>
+        [doc] book/second-edition (book) <host>
+        [doc] book/2018-edition (book) <host>
+        [build] rustdoc 1 <host>
+        [doc] rustc 1 <host> -> standalone 2 <host>
+        [doc] rustc 1 <host> -> std 1 <host> crates=[alloc,compiler_builtins,core,panic_abort,panic_unwind,proc_macro,rustc-std-workspace-core,std,std_detect,sysroot,test,unwind]
+        [doc] rustc 1 <host> -> rustc 2 <host>
+        [build] rustc 1 <host> -> rustc 2 <host>
+        [doc] rustc 1 <host> -> Rustdoc 2 <host>
+        [doc] rustc 1 <host> -> Rustfmt 2 <host>
+        [build] rustc 1 <host> -> error-index 2 <host>
+        [doc] rustc 1 <host> -> error-index 2 <host>
+        [doc] nomicon (book) <host>
+        [doc] rustc 1 <host> -> reference (book) 2 <host>
+        [doc] rustdoc (book) <host>
+        [doc] rust-by-example (book) <host>
+        [build] rustc 0 <host> -> LintDocs 1 <host>
+        [doc] rustc (book) <host>
+        [doc] rustc 1 <host> -> Cargo 2 <host>
+        [doc] cargo (book) <host>
+        [doc] rustc 1 <host> -> Clippy 2 <host>
+        [doc] clippy (book) <host>
+        [doc] rustc 1 <host> -> Miri 2 <host>
+        [doc] embedded-book (book) <host>
+        [doc] edition-guide (book) <host>
+        [doc] style-guide (book) <host>
+        [build] rustdoc 0 <host>
+        [doc] rustc 0 <host> -> Tidy 1 <host>
+        [doc] rustc 0 <host> -> Bootstrap 1 <host>
+        [doc] rustc 1 <host> -> releases 2 <host>
+        [doc] rustc 0 <host> -> RunMakeSupport 1 <host>
+        [doc] rustc 0 <host> -> BuildHelper 1 <host>
+        [doc] rustc 0 <host> -> Compiletest 1 <host>
+        [build] rustc 0 <host> -> RustInstaller 1 <host>
+        "
+        );
+    }
+
+    #[test]
     fn dist_extended() {
         let ctx = TestCtx::new();
         insta::assert_snapshot!(
@@ -2473,7 +2526,6 @@ mod snapshot {
         [doc] rustc 0 <host> -> Cargo 1 <host>
         ");
     }
-
     #[test]
     fn doc_cargo_stage_2() {
         let ctx = TestCtx::new();

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -2463,11 +2463,24 @@ mod snapshot {
     }
 
     #[test]
-    fn doc_cargo() {
+    fn doc_cargo_stage_1() {
         let ctx = TestCtx::new();
         insta::assert_snapshot!(
             ctx.config("doc")
                 .path("cargo")
+                .render_steps(), @r"
+        [build] rustdoc 0 <host>
+        [doc] rustc 0 <host> -> Cargo 1 <host>
+        ");
+    }
+
+    #[test]
+    fn doc_cargo_stage_2() {
+        let ctx = TestCtx::new();
+        insta::assert_snapshot!(
+            ctx.config("doc")
+                .path("cargo")
+                .stage(2)
                 .render_steps(), @r"
         [build] llvm <host>
         [build] rustc 0 <host> -> rustc 1 <host>

